lib/inc/drogon/ThreadStorage.h

```diff
@@ -0,0 +1,155 @@
+/**
+ *
+ *  ThreadStorage.h
+ *  Daniel Mensinger
+ *
+ *  Copyright 2019, Daniel Mensinger.  All rights reserved.
+ *  https://github.com/an-tao/drogon
+ *  Use of this source code is governed by a MIT license
+ *  that can be found in the License file.
+ *
+ *  Drogon
+ *
+ */
+
+#pragma once
+
+#include <memory>
+#include <vector>
+#include <limits>
+
+#include <drogon/HttpAppFramework.h>
+
+namespace drogon
+{
+/**
+ * @brief Base class for all thread storage containers
+ */
+class ThreadDataBase
+{
+  public:
+    // Requierd for std::dynamic_pointer_cast
+    virtual ~ThreadDataBase()
+    {
+    }
+};
+
+using ThreadDataPtr = std::shared_ptr<ThreadDataBase>;
+
+/**
+ * @brief Abstract Utility class for thread storage handling
+ *
+ * Thread storage allows the efficient handling of reusable data without thread
+ * synchronisation. For instance, such a thread storage would be useful to store
+ * database connections.
+ *
+ * @note initializeThreadStorage() MUST be called in the constructor of the
+ * derived class
+ *
+ * Example usage:
+ *
+ * @code
+ * class MyThreadData : public ThreadDataBase {
+ *   public:
+ *     int threadLocal;
+ * };
+ *
+ * class MyController : public HttpController<MyController>,
+ *                      public ThreadStorage {
+ *   public:
+ *      METHOD_LIST_BEGIN
+ *      ADD_METHOD_TO(MyController::endpoint, "/some/path", Get);
+ *      METHOD_LIST_END
+ *
+ *      MyController()
+ *      {
+ *          initThreadStorage();
+ *      }
+ *
+ *      ThreadDataPtr createDataContainer(size_t thread) override
+ *      {
+ *          return std::make_shared<MyThreadData>();
+ *      }
+ *
+ *      void login(const HttpRequestPtr &req,
+ *                 std::function<void (const HttpResponsePtr &)> &&callback) {
+ *          auto data = getThreadData<MyThreadData>();
+ *
+ *          // handle the request
+ *      }
+ * };
+ * @endcode
+ */
+class ThreadStorage
```

Comment:
How about using:
```
template<typename DataType>
class ThreadStorage
{... 
    ThreadStorage()
    {
        size_t numThreads = app().getThreadNum();
        assert(numThreads > 0 &&
               numThreads != std::numeric_limits<size_t>::max());
        _storage.resize(numThreads);
    }
   shared_ptr<DataType> &getThreadData() //returning reference allows user to set it.
   {
...
   }
    std::vector<shared_ptr<DataType>> _storage;
.....
}
```
this solution doesn't need the ThreadDataBase class, and the dynamic_pointer_cast<>(the cost of this is not little).

And users can just use it in controllers instead of inheriting.

```
class MyController : public HttpController<MyController>
{
   public:
       METHOD_LIST_BEGIN
       ADD_METHOD_TO(MyController::endpoint, "/some/path", Get);
       METHOD_LIST_END
 
       ThreadStorage<int> _storage1;
       ThreadStorage<DbClient> _storage2;
       ThreadStorage<MongoDBConnection> _storage3;
 
       void login(const HttpRequestPtr &req,
                  std::function<void (const HttpResponsePtr &)> &&callback) {
           auto data = getThreadData<MyThreadData>();
 
          // handle the request
       }
  };
```